module/VuFind/src/VuFind/Controller/ApiTrait.php

```diff
@@ -0,0 +1,732 @@
+            'description' => 'Statement of responsibility that goes with the title',
+            'type' => 'string'
+        ],
+        'toc' => [
+            'method' => 'getTOC',
+            'description' => 'Table of contents',
+            'type' => 'array',
+            'items' => [
+                'type' => 'string'
+            ]
+        ],
+        'urls' => [
+            'method' => 'getRecordURLs',
+            'description' => 'URLs contained in the record',
+            'type' => 'array',
+            'items' => [
+                '$ref' => '#/definitions/Url'
+            ]
+        ]
+    ];
+
+    /**
+     * Default record fields to return if a request does not define the fields
+     *
+     * @var array
+     */
+    protected $defaultRecordFields = [
+        'authors',
+        'formats',
+        'id',
+        'languages',
+        'rating',
+        'series',
+        'subjects',
+        'title',
+        'urls'
+    ];
+
+    /**
+     * Execute the request
+     *
+     * @param \Zend\Mvc\MvcEvent $e Event
+     *
+     * @return mixed
+     * @throws Exception\DomainException
+     */
+    public function onDispatch(\Zend\Mvc\MvcEvent $e)
+    {
+        // Add CORS headers and handle OPTIONS requests. This is a simplistic
+        // approach since we allow any origin. For more complete CORS handling
+        // a module like zfr-cors could be used.
+        $response = $this->getResponse();
+        $headers = $response->getHeaders();
+        $headers->addHeaderLine('Access-Control-Allow-Origin: *');
+        $request = $this->getRequest();
+        if ($request->getMethod() == 'OPTIONS') {
+            // Disable session writes
+            $this->disableSessionWrites();
+            $headers->addHeaderLine(
+                'Access-Control-Allow-Methods', 'GET, POST, OPTIONS'
+            );
+            $headers->addHeaderLine('Access-Control-Max-Age', '86400');
+
+            return $this->output(null, 204);
+        }
+        return parent::onDispatch($e);
+    }
+
+    /**
+     * Index action
+     *
+     * @return \Zend\Http\Response
+     */
+    public function indexAction()
+    {
+        // Disable session writes
+        $this->disableSessionWrites();
+
+        if (null === $this->getRequest()->getQuery('swagger')) {
+            $urlHelper = $this->getViewRenderer()->plugin('url');
+            $base = rtrim($urlHelper('home'), '/');
+            $url = "$base/swagger-ui/index.html?url="
+                . urlencode("$base/api?swagger");
+            return $this->redirect()->toUrl($url);
+        }
+        $response = $this->getResponse();
+        $headers = $response->getHeaders();
+        $headers->addHeaderLine('Content-type', 'application/json');
+        $config = $this->getConfig();
+        $results = $this->getResultsManager()->get($this->searchClassId);
+        $options = $results->getOptions();
+        $searchConfig = $this->getConfig($options->getSearchIni());
```

Comment:
Is there a compelling reason to read raw data out of the .ini files rather than making use of the processed data from the Options object? There are some cases where the PHP code may modify or extend the configuration settings, and those options will be lost with this approach. (This may or may not be important for the purposes of the API -- but it seems safer/more consistent to use the more abstract interfaces rather than raw data whenever practical).